stdlib/_ast.pyi

```diff
@@ -1,61 +1,81 @@
 import sys
 import typing_extensions
-from typing import Any, ClassVar, Literal
+from typing import Any, ClassVar, Generic, Literal, TypedDict, overload
+from typing_extensions import Unpack
 
 PyCF_ONLY_AST: Literal[1024]
 PyCF_TYPE_COMMENTS: Literal[4096]
 PyCF_ALLOW_TOP_LEVEL_AWAIT: Literal[8192]
 
+# Used for node end positions in constructor keyword arguments
+_EndPositionT = typing_extensions.TypeVar("_EndPositionT", int, int | None, default=int | None)  # noqa: Y023
+
 # Alias used for fields that must always be valid identifiers
 # A string `x` counts as a valid identifier if both the following are True
 # (1) `x.isidentifier()` evaluates to `True`
 # (2) `keyword.iskeyword(x)` evaluates to `False`
 _Identifier: typing_extensions.TypeAlias = str
 
+# Corresponds to the names in the `_attributes` class variable which is non-empty in certain AST nodes
+class _Attributes(TypedDict, Generic[_EndPositionT], total=False):
+    lineno: int
+    col_offset: int
+    end_lineno: _EndPositionT
+    end_col_offset: _EndPositionT
+
+if sys.version_info >= (3, 9):
+    _SliceAttributes: typing_extensions.TypeAlias = _Attributes
+else:
+    class _SliceAttributes(TypedDict): ...
+
 class AST:
     if sys.version_info >= (3, 10):
         __match_args__ = ()
     _attributes: ClassVar[tuple[str, ...]]
     _fields: ClassVar[tuple[str, ...]]
-    def __init__(self, *args: Any, **kwargs: Any) -> None: ...
-    # TODO: Not all nodes have all of the following attributes
-    lineno: int
-    col_offset: int
-    end_lineno: int | None
-    end_col_offset: int | None
-    type_comment: str | None
 
 class mod(AST): ...
 class type_ignore(AST): ...
 
 class TypeIgnore(type_ignore):
     if sys.version_info >= (3, 10):
         __match_args__ = ("lineno", "tag")
+    lineno: int
     tag: str
+    def __init__(self, lineno: int, tag: str) -> None: ...
 
 class FunctionType(mod):
     if sys.version_info >= (3, 10):
         __match_args__ = ("argtypes", "returns")
     argtypes: list[expr]
     returns: expr
+    def __init__(self, argtypes: list[expr], returns: expr) -> None: ...
 
 class Module(mod):
     if sys.version_info >= (3, 10):
         __match_args__ = ("body", "type_ignores")
     body: list[stmt]
     type_ignores: list[TypeIgnore]
+    def __init__(self, body: list[stmt], type_ignores: list[TypeIgnore]) -> None: ...
```

Comment:
These list fields should default to the empty list on 3.13+.

Reply:
Addressed in https://github.com/python/typeshed/pull/11880/commits/c82a5fa6fd2a9c57321e2685128df23bcdcfe289